File: app/src/main/java/tech/blur/bookcrossing/features/BaseActivity.java
package tech.blur.bookcrossing.features;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import javax.inject.Inject;

import ru.terrakok.cicerone.NavigatorHolder;
import tech.blur.bookcrossing.features.core.GlobalNavigator;
import tech.blur.bookcrossing.features.core.qualifier.Global;

public abstract class BaseActivity extends AppCompatActivity {

    @Inject
    @Global
    NavigatorHolder globalNavigatorHolder;

    protected abstract <T extends MvpView> MvpPresenter<T> getPresenter();

    protected abstract <T extends MvpView> T getMvpView();

    private MvpPresenter<MvpView> presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = getPresenter();
    }

    @Override
    protected void onStart() {
        super.onStart();
        presenter.attachView(getMvpView());
    }

    @Override
    protected void onStop() {
        super.onStop();
        presenter.detachView();
    }

    @Override
    protected void onResume() {
        super.onResume();
        globalNavigatorHolder.setNavigator(new GlobalNavigator(this));
    }
}

File: app/src/main/java/tech/blur/bookcrossing/features/BaseActivity.kt
package tech.blur.bookcrossing.features

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

abstract class BaseActivity : AppCompatActivity() {

    private var presenter: MvpPresenter<MvpView>? = null

    protected abstract fun <T : MvpView> getPresenter(): MvpPresenter<T>

    protected abstract fun <T : MvpView> getMvpView(): T

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = getPresenter()
    }

    override fun onStart() {
        super.onStart()
        presenter!!.attachView(getMvpView())
    }

    override fun onStop() {
        super.onStop()
        presenter!!.detachView()
    }
}
